lib/common/socket.c

```diff
@@ -992,17 +976,27 @@ void h2o_socket_ssl_handshake(h2o_socket_t *sock, SSL_CTX *ssl_ctx, const char *
             h2o_socket_read_start(sock, proceed_handshake);
     } else {
         if (session_cache != NULL) {
-            sock->ssl->handshake.client.session_cache = session_cache;
-            h2o_iovec_t session_cache_key = get_ssl_session_cache_key(sock, server_name);
-            sock->ssl->handshake.client.session_cache_key = session_cache_key;
-            sock->ssl->handshake.client.session_cache_key_hash = h2o_cache_calchash(session_cache_key.base, session_cache_key.len);
-            fprintf(stderr, "@@@ session_cache_key: %s\n", session_cache_key.base); // FIXME: nagata
-
-            h2o_cache_ref_t *cacheref =
-                h2o_cache_fetch(session_cache, h2o_now(h2o_socket_get_loop(sock)), sock->ssl->handshake.client.session_cache_key,
-                                sock->ssl->handshake.client.session_cache_key_hash);
-            if (cacheref != NULL) {
-                SSL_set_session(sock->ssl->ssl, (SSL_SESSION *)cacheref->value.base);
+            struct sockaddr sa;
+            uint16_t port;
+            if (h2o_socket_getpeername(sock, &sa) != 0 && (port = h2o_socket_getport(&sa)) > 0) {
+                /* session cache is available */
+                h2o_iovec_t session_cache_key;
+                size_t len = strlen(server_name) + sizeof(H2O_UINT16_LONGEST_STR) + 1;
+                session_cache_key.base = h2o_mem_alloc(len + 1);
+                session_cache_key.len = snprintf(session_cache_key.base, len + 1, "%s:%" PRIu16, server_name, port);
```

Comment:
Let's just use `sprintf`, considering the fact that we are sure the generated string fits in memory.